File: Python/algorithms/ford_fulkerson_algorithm/ford_fulkerson.py
class FordFulkerson:
    def __init__(self, V):
        """
        Constructor for the FordFulkerson class.

        :param V: The number of vertices in the flow network.
        """
        self.V = V
        self.capacity = [[0] * V for _ in range(V)]
        self.residual_capacity = [[0] * V for _ in range(V)]
        self.visited = [False] * V
        self.parent = [-1] * V

    def addEdge(self, u, v, capacity):
        """
        Add an edge with capacity to the flow network.

        :param u: The source vertex of the edge.
        :param v: The target vertex of the edge.
        :param capacity: The capacity of the edge.
        """
        self.capacity[u][v] = capacity

    def bfs(self, source, drain):
        """
        Perform Breadth-First Search (BFS) to find augmenting paths in the residual graph.

        :param source: The source vertex.
        :param drain: The sink (target) vertex.
        :return: True if there is an augmenting path, False otherwise.
        """
        self.visited = [False] * self.V
        queue = [source]
        self.visited[source] = True
        self.parent[source] = -1

        while queue:
            u = queue.pop(0)
            for v in range(self.V):
                if not self.visited[v] and self.residual_capacity[u][v] > 0:
                    queue.append(v)
                    self.parent[v] = u
                    self.visited[v] = True

        return self.visited[drain]

    def minCapacityOnPath(self, u, v):
        """
        Find the minimum capacity on the path from u to v in the residual graph.

        :param u: The source vertex.
        :param v: The target vertex.
        :return: The minimum capacity on the path.
        """
        min_capacity = float("inf")
        current = v
        while current != u:
            previous = self.parent[current]
            min_capacity = min(min_capacity, self.residual_capacity[previous][current])
            current = previous
        return min_capacity

    def updateResidualCapacity(self, u, v, flow):
        """
        Update the residual capacities in the residual graph after augmenting flow.

        :param u: The source vertex.
        :param v: The target vertex.
        :param flow: The flow being added to the edge (u, v).
        """
        self.residual_capacity[u][v] -= flow
        self.residual_capacity[v][u] += flow

    def maxFlow(self, source, drain):
        """
        Find the maximum flow in the flow network using the Ford-Fulkerson algorithm.

        :param source: The source vertex.
        :param drain: The sink (target) vertex.
        :return: The maximum flow value.
        """
        # Initialize residual capacity
        for u in range(self.V):
            for v in range(self.V):
                self.residual_capacity[u][v] = self.capacity[u][v]

        max_flow = 0
        while self.bfs(source, drain):
            flow = self.minCapacityOnPath(source, drain)
            max_flow += flow
            current = drain
            while current != source:
                previous = self.parent[current]
                self.updateResidualCapacity(previous, current, flow)
                current = previous

        return max_flow

File: C++/ford_fulkerson/ford_fulkerson.cpp
//
// Created by hunor on 21.08.2023.
//

#include "ford_fulkerson.h"

FordFulkerson::FordFulkerson(int V) : V(V) {
    this->capacity.resize(V, std::vector<int>(V, 0));
    residualCapacity.resize(V, std::vector<int>(V, 0));
    visited.resize(V, false);
    parent.resize(V, -1);
}

void FordFulkerson::addEdge(int u, int v, int capacity) {
    this->capacity[u][v] = capacity;
}

bool FordFulkerson::bfs(int source, int drain) {
    std::fill(visited.begin(), visited.end(), false);
    std::queue<int> q;
    q.push(source);
    visited[source] = true;
    parent[source] = -1;

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        for (int v = 0; v < V; ++v) {
            if (!visited[v] && residualCapacity[u][v] > 0) {
                q.push(v);
                parent[v] = u;
                visited[v] = true;
            }
        }
    }

    return visited[drain];
}

int FordFulkerson::minCapacityOnPath(int u, int v) {
    int minCapacity = INT_MAX;
    int current = v;
    while (current != u) {
        int previous = parent[current];
        minCapacity = std::min(minCapacity, residualCapacity[previous][current]);
        current = previous;
    }
    return minCapacity;
}

void FordFulkerson::updateResidualCapacity(int u, int v, int flow) {
    residualCapacity[u][v] -= flow;
    residualCapacity[v][u] += flow;
}

int FordFulkerson::maxFlow(int source, int drain) {
    // Initialize residual capacity
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
            residualCapacity[u][v] = capacity[u][v];
        }
    }

    int maxFlow = 0;
    while (bfs(source, drain)) {
        int flow = minCapacityOnPath(source, drain);
        maxFlow += flow;
        int current = drain;
        while (current != source) {
            int previous = parent[current];
            updateResidualCapacity(previous, current, flow);
            current = previous;
        }
    }

    return maxFlow;
}
